test_quality_of_service/test/test_deadline.cpp

```
@@ -0,0 +1,121 @@
+  int expected_number_of_pub_events = 5;
+  const std::chrono::milliseconds deadline_duration = 1s;
+  std::tuple<size_t, size_t> deadline_duration_tuple = chrono_milliseconds_to_size_t(
+    deadline_duration);
+  const std::chrono::milliseconds max_test_length = 10s;
+  const std::chrono::milliseconds publish_rate = deadline_duration / 5;
+
+  // used for lambda capture
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec = executor;
+  int total_number_of_publisher_deadline_events = 0;
+  int total_number_of_subscriber_deadline_events = 0;
+  int last_pub_count = 0;
+  int last_sub_count = 0;
+
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_default;
+  qos_profile.deadline.sec = std::get<0>(deadline_duration_tuple);
+  qos_profile.deadline.nsec = std::get<1>(deadline_duration_tuple);
+
+  // setup subscription options and callback
+  rclcpp::SubscriptionOptions<> subscriber_options;
+  subscriber_options.qos_profile = qos_profile;
+  subscriber_options.event_callbacks.deadline_callback =
+    [&last_pub_count,
+    &total_number_of_subscriber_deadline_events](rclcpp::QOSDeadlineRequestedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSDeadlineRequestedInfo callback");
+      total_number_of_subscriber_deadline_events++;
+      // assert the correct value on a change
+      ASSERT_EQ(1, event.total_count_change);
+      last_pub_count = event.total_count;
+    };
+
+  // setup publishing options and callback
+  rclcpp::PublisherOptions<> publisher_options;
+  publisher_options.qos_profile = qos_profile;
+  publisher_options.event_callbacks.deadline_callback =
+    [&last_sub_count,
+    &total_number_of_publisher_deadline_events](rclcpp::QOSDeadlineOfferedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSDeadlineOfferedInfo callback");
+      total_number_of_publisher_deadline_events++;
+      // assert the correct value on a change
+      ASSERT_EQ(1, event.total_count_change);
```

Assert total_count against our locally-tracked counter `total_number_of_publisher_deadline_events`?

---

This is a simple check to ensure the callback is working (to fail fast). The total count is checked at the end.